### ISSUE TITLE
Rule#each. undefined method `map' for nil:NilClass

### ISSUE THREAD

**Issue by yuuan** [dry-rb/dry-validation#693]
## Describe the bug

NoMethodError was raised when maybe array rule receives nil.

## To Reproduce

```ruby
class Sample < Dry::Validation::Contract
  params do
    required(:foo).maybe(:array).each(:int?)
  end
end

Sample.new.call(foo: nil)
```

## Expected behavior

Don't run each if the rule with maybe receives nil.

## My environment

- Affects my production application: **NO**
- Ruby version: ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]
- OS: Fedora release 35 (Thirty Five)

### Stack trace
```
 NoMethodError:
   undefined method `map' for nil:NilClass
 # /usr/local/bundle/gems/dry-logic-1.2.0/lib/dry/logic/operations/each.rb:15:in `call'
 # /usr/local/bundle/gems/dry-logic-1.2.0/lib/dry/logic/operations/and.rb:26:in `call'
 # /usr/local/bundle/gems/dry-logic-1.2.0/lib/dry/logic/operations/key.rb:41:in `call'
 # /usr/local/bundle/gems/dry-logic-1.2.0/lib/dry/logic/operations/implication.rb:22:in `call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/rule_applier.rb:35:in `block in call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/rule_applier.rb:32:in `each'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/rule_applier.rb:32:in `call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/step.rb:35:in `call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/processor_steps.rb:40:in `block in call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/processor_steps.rb:38:in `each'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/processor_steps.rb:38:in `call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/processor.rb:88:in `block in call'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/result.rb:37:in `new'
 # /usr/local/bundle/gems/dry-schema-1.8.0/lib/dry/schema/processor.rb:87:in `call'
 # /usr/local/bundle/gems/dry-validation-1.7.0/lib/dry/validation/contract.rb:105:in `call'
```

### Gemfile.lock
```gemfile
GEM
  remote: https://rubygems.org/
  specs:
    concurrent-ruby (1.1.9)
    dry-configurable (0.13.0)
      concurrent-ruby (~> 1.0)
      dry-core (~> 0.6)
    dry-container (0.9.0)
      concurrent-ruby (~> 1.0)
      dry-configurable (~> 0.13, >= 0.13.0)
    dry-core (0.7.1)
      concurrent-ruby (~> 1.0)
    dry-inflector (0.2.1)
    dry-initializer (3.0.4)
    dry-logic (1.2.0)
      concurrent-ruby (~> 1.0)
      dry-core (~> 0.5, >= 0.5)
    dry-schema (1.8.0)
      concurrent-ruby (~> 1.0)
      dry-configurable (~> 0.13, >= 0.13.0)
      dry-core (~> 0.5, >= 0.5)
      dry-initializer (~> 3.0)
      dry-logic (~> 1.0)
      dry-types (~> 1.5)
    dry-types (1.5.1)
      concurrent-ruby (~> 1.0)
      dry-container (~> 0.3)
      dry-core (~> 0.5, >= 0.5)
      dry-inflector (~> 0.1, >= 0.1.2)
      dry-logic (~> 1.0, >= 1.0.2)
    dry-validation (1.7.0)
      concurrent-ruby (~> 1.0)
      dry-container (~> 0.7, >= 0.7.1)
      dry-core (~> 0.5, >= 0.5)
      dry-initializer (~> 3.0)
      dry-schema (~> 1.8, >= 1.8.0)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  dry-validation

RUBY VERSION
   ruby 3.0.2p107

BUNDLED WITH
   2.2.28
```

**Comment by flash-gordon** [comment]
It's incorrect, calling `.each` just adds another check, it's not connected to the `nil`-check added by `maybe`.

Change it to
```ruby
class Sample < Dry::Validation::Contract
  params do
    required(:foo).maybe(array[:integer])
  end
end
```

Or this one seems to work as well
```ruby
class Sample < Dry::Validation::Contract
  params do
    required(:foo).maybe do
      array(:integer)
    end
  end
end
```
